nntrainer/tensor/cuda_operations/gemm_int4_cpu.cpp

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2025 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file   gemm_int4_cpu.cpp
+ * @brief  CPU implementation of int4 GEMM operation
+ * @author Samsung Electronics Co., Ltd.
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include "gemm_int4_cpu.h"
+#include "fp16.h"
+#include "int4_utils.h"
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+namespace nntrainer {
+
+void gemm_int4_cpu(const void *input, const void *weights, const void *scales,
+                   const void *input_scales, float *output, unsigned int M,
+                   unsigned int N, unsigned int K,
+                   unsigned int quantization_group_size) {
+
+  const int8_t *q_input = static_cast<const int8_t *>(input);
+  const uint8_t *q_weights = static_cast<const uint8_t *>(weights);
+  const uint16_t *w_scales = static_cast<const uint16_t *>(scales);
+  const uint16_t *in_scales = static_cast<const uint16_t *>(input_scales);
+
+  // Buffer to hold one dequantized row of weights (size K)
+  // Since weights are packed as N rows x K columns (passed to quantize),
+  // the n-th row of weights corresponds to the n-th column of the GEMM matrix
+  // B.
+  std::vector<float> dequantized_weight_row(K);
+
+  // Input quantization layout parameters
+  // Matches cpu_quantize_input_int8_pad logic
+  unsigned int alignK = (K + quantization_group_size - 1) /
+                        quantization_group_size * quantization_group_size;
+  unsigned int groups_in_row = alignK / quantization_group_size;
+
+  // Iterate over columns of output (N)
+  for (unsigned int n = 0; n < N; ++n) {
+    // 1. Dequantize the n-th row of weights (which is n-th col of B)
+    Int4Utils::dequantizePackedRow(
+      const_cast<uint8_t *>(q_weights), const_cast<uint16_t *>(w_scales), N, K,
+      quantization_group_size, n, dequantized_weight_row.data());
+
+    // 2. Compute dot product with each row of input (M)
+    for (unsigned int m = 0; m < M; ++m) {
+      float sum = 0.0f;
```

Comment:
1. Please check if having loop-m inside loop-k would be faster than loop-k inside loop-m or not.
2. Try to cache caculation as much as possible. (looks like the same calculation/memory-op is repeated in the loop)